### PR TITLE
Stats: Tables: `statusHistory` introduction

### PR DESCRIPTION
**Pull request by kamsi**
Needed for:

https://docs.google.com/document/d/1i92UhTr_krhLYVc9TA88WGIrvwPAnM9x84WHCFpVtQc/edit

In order to be able to get all the information (like how many pendings were there on a given day) for the new stats we need to gather some more data. I propose to create `ProcessingStep.prototype.flowLog` as a nested map. It's entries must contain the following:

`status` - Status that was setup for the step.
`stamp` - Moment when given status change was triggered.

The setup should be performed from within the current flow service: https://github.com/egovernment/eregistrations/blob/master/server/services/business-process-flow.js

All the status changes should be written to the `flowLog` of corresponding steps.

### CODE REVIEW

**Comment by kamsi** [comment]
@medikoo let's have this approved before I test it in all systems, and prepare adjusts (only thing needed should be initialization in master process).

File: model/lib/status-history-log.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+var memoize          = require('memoizee/plain')
+  , defineStringLine = require('dbjs-ext/string/string-line');
+
+module.exports = memoize(function (db) {
+	var StringLine = defineStringLine(db);
+
+	return db.Object.extend('StatusHistoryLog', {
```
Comment (medikoo) [inline model/lib/status-history-log.js]: Maybe let's name it `StatusHistoryItem`, it should not be confused with `statusLog` which is similar but different

File: server/services/status-history-logger.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  , Set                           = require('es6-set')
+  , processingStepsMeta           = require('../../processing-steps-meta')
+  , uuid                          = require('time-uuid')
+  , uniqIdPrefix                  = 'abcdefghiklmnopqrstuvxyz'[Math.floor(Math.random() * 24)];
+
+var storeLog = function (storage, logPath) {
+	return function (event) {
+		var status, oldStatus;
+
+		if (event.type !== 'computed' || !event.old) return;
```
Comment (medikoo) [inline server/services/status-history-logger.js]: We shouldn't discard record if it's first record in history for this property

File: test/model/lib/status-history-log.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// TODO
+
+'use strict';
```
Comment (medikoo) [inline test/model/lib/status-history-log.js]: Let's add some sanity test

**Comment by medikoo** [comment]
@kamsi conflicts to be resolved

File: server/services/status-history-logger.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+var storeLog = function (storage, logPath/*, options */) {
+	var options = Object(arguments[2]);
+	return function (event) {
+		var status, oldStatus, resolvedPath, logPathResolved;
+		logPathResolved = logPath ? logPath + '/' + getPathSuffix() : getPathSuffix();
+
+		if (event.type !== 'computed') return;
+		status    = unserializeValue(event.data.value) || null;
+		oldStatus = (event.old && unserializeValue(event.old.value)) || null;
```
Comment (medikoo) [inline server/services/status-history-logger.js]: As we discussed let's not store `null` but `''` (`null` is not acceptable value for `cardinalPropertyKey`)

File: server/services/status-history-logger.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+		// We ignore such cases, they may happen when direct overwrites computed
+		if (status === oldStatus) return;
+		if (options.processorPath) {
+			storage.get(event.ownerId + '/' + options.processorPath)(function (data) {
+				if (data && data.value[0] === '7') {
+					resolvedPath = event.ownerId + '/' + logPathResolved + '/processor';
+					storage.store(resolvedPath, data.value);
+					debug('Stored processor %s for the path %s', data.value, resolvedPath);
```
Comment (medikoo) [inline server/services/status-history-logger.js]: No need for those logs, as db logs, show all stores

File: server/services/status-history-logger.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+				}
+			});
+		}
+		resolvedPath = event.ownerId + '/' + logPathResolved + '/status';
+		storage.store(resolvedPath, serializeValue(status));
+		debug('Stored status %s for the path %s', status, resolvedPath);
```
Comment (medikoo) [inline server/services/status-history-logger.js]: Same here

File: server/services/status-history-logger.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		status    = unserializeValue(event.data.value) || null;
+		oldStatus = (event.old && unserializeValue(event.old.value)) || null;
+
+		// We ignore such cases, they may happen when direct overwrites computed
+		if (status === oldStatus) return;
+		if (options.processorPath) {
+			storage.get(event.ownerId + '/' + options.processorPath)(function (data) {
+				if (data && data.value[0] === '7') {
+					resolvedPath = event.ownerId + '/' + logPathResolved + '/processor';
+					storage.store(resolvedPath, data.value);
```
Comment (medikoo) [inline server/services/status-history-logger.js]: This returns promise which should be handled

File: server/services/status-history-logger.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+					debug('Stored processor %s for the path %s', data.value, resolvedPath);
+				}
+			});
```
Comment (medikoo) [inline server/services/status-history-logger.js]: Returns promise that should be handled

File: server/services/status-history-logger.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+		}
+		resolvedPath = event.ownerId + '/' + logPathResolved + '/status';
+		storage.store(resolvedPath, serializeValue(status));
```
Comment (medikoo) [inline server/services/status-history-logger.js]: Returns promise that should be handled

File: model/business-process-new/flow.js [hunk 10]
```diff
@@ -109,12 +109,32 @@ module.exports = memoize(function (db/*, options*/) {
...
 			if (this.isRejected) return 'rejected';
 			if (this.isClosed) return 'closed';
 
+			_observe(this.processingSteps.applicable).some(function (step) {
+				if (_observe(step._status) === 'rejected') {
+					result = step.status;
+					return true;
+				}
+			});
+			if (result) return result;
```
Comment (medikoo) [inline model/business-process-new/flow.js]: Looks a bit overcomplicated. It seams that cleaner would be:

```javascript
if (_observe(this.processingSteps.applicable).some(function (step) {
    return _observe(step._status) === 'rejected';
  })) {
  return 'rejected';
}
```

File: test/model/lib/status-history-item.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+// TODO
```
Comment (medikoo) [inline test/model/lib/status-history-item.js]: Let's remove that comment

**Pull request merged**